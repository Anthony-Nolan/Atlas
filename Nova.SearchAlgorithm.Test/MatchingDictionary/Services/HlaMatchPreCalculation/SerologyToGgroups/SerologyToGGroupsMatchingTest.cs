using ApprovalTests;
using ApprovalTests.Reporters;
using Nova.SearchAlgorithm.MatchingDictionary.Models.HLATypings;
using Nova.SearchAlgorithm.MatchingDictionary.Models.MatchingTypings;
using NUnit.Framework;
using System.Collections.Generic;
using System.Linq;

namespace Nova.SearchAlgorithm.Test.MatchingDictionary.Services.HlaMatchPreCalculation.SerologyToGgroups
{
    [TestFixtureSource(typeof(MatchedHlaTestFixtureArgs), nameof(MatchedHlaTestFixtureArgs.MatchedSerologies))]
    [UseReporter(typeof(NUnitReporter))]
    [ApprovalTests.Namers.UseApprovalSubdirectory("../../../../Resources/MDPreCalc")]
    public class SerologyToGGroupsMatchingTest : MatchedOnTestBase<MatchedSerology>
    {
        public SerologyToGGroupsMatchingTest(IEnumerable<MatchedSerology> matchingTypes) : base(matchingTypes)
        {
        }

        [Test]
        public void ValidSerologyHaveAtLeastOneGGroup()
        {
            var gGroupCounts = MatchedHlaTypings
                .Where(m => !m.HlaTyping.IsDeleted && m.HlaTyping is SerologyTyping)
                .Select(m => new { HlaType = m.HlaTyping, GGroupCount = m.MatchingGGroups.Count() })
                .ToList();

            Assert.IsTrue(gGroupCounts.All(g => g.GGroupCount > 0));
        }

        [Test]
        public void BroadWhereSplitHasAssociatedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.A, "9"));
        }

        [Test]
        public void BroadWhereSplitHasNoAssociatedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.Dqb1, "1"));
        }

        [Test]
        public void BroadHasSplitAndAssociatedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "21"));
        }

        [Test]
        public void SplitHasAssociatedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "51"));
        }

        [Test]
        public void SplitNoAssociatedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.C, "10"));
        }

        [Test]
        public void AssociatedWithSplitGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "3902"));
        }

        [Test]
        public void AssociatedWithBroadGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "4005"));
        }

        [Test]
        public void AssociatedWithNotSplitGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.Drb1, "103"));
        }

        [Test]
        public void NotSplitWithAssociatedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.A, "2"));
        }

        [Test]
        public void NotSplitNoAssociatedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.A, "1"));
        }

        [Test]
        public void DeletedGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.C, "11"));
        }

        [Test]
        public void B15BroadGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "15"));
        }

        [Test]
        public void B15SplitGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "62"));
        }

        [Test]
        public void B70BroadGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "70"));
        }

        [Test]
        public void B70SplitGGroupsAreCorrect()
        {
            Approvals.Verify(GetGGroupsAsString(MatchLocus.B, "72"));
        }

        [Test]
        public void B15And70ShareMatchingGGroups()
        {
            var b15 = GetSingleMatchingTyping(MatchLocus.B, "15").MatchingGGroups;
            var b70 = GetSingleMatchingTyping(MatchLocus.B, "70").MatchingGGroups;

            Approvals.Verify(string.Join("\r\n", b15.Intersect(b70).OrderBy(p => p)));
        }

        private string GetGGroupsAsString(MatchLocus matchLocus, string serologyName)
        {
            return string.Join("\r\n",
                GetSingleMatchingTyping(matchLocus, serologyName)
                .MatchingGGroups
                .OrderBy(p => p));
        }
    }
}
