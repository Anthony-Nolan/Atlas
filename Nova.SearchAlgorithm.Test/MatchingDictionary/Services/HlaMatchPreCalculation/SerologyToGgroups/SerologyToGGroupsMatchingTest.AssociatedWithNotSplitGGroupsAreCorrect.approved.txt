01:01:01G
01:01:02
01:01:03
01:01:04
01:01:05
01:01:06
01:01:07
01:01:08
01:01:09
01:01:10
01:01:11
01:01:12
01:01:13
01:01:14
01:01:15
01:01:16
01:01:17
01:01:18
01:01:19
01:01:20
01:01:21
01:01:22
01:01:23
01:01:24
01:01:25
01:01:26
01:01:27
01:01:28
01:01:29
01:01:30
01:02:01G
01:02:02
01:02:03
01:02:04
01:02:05
01:02:06
01:02:07
01:02:08
01:02:09
01:02:10
01:02:11
01:02:12
01:03:01
01:03:02
01:04
01:05
01:06
01:07
01:08
01:09
01:10
01:11:01
01:11:02
01:12
01:13
01:14
01:15
01:16
01:17
01:18:01
01:18:02
01:19
01:20:01
01:20:02
01:21
01:22
01:23
01:24:01
01:24:02
01:25
01:26
01:27
01:28
01:29:01
01:29:02
01:30
01:31
01:32
01:34
01:35
01:36
01:37
01:38
01:41
01:42
01:43
01:44:01
01:44:02
01:45
01:46
01:47
01:48
01:49
01:51
01:53
01:54
01:55
01:56
01:57
01:58
01:59
01:60
01:61
01:63
01:64
01:65:01
01:65:02
01:66
01:69
01:70
01:71
01:72
01:73
01:74
01:75
01:76
01:78
01:79
01:80
01:81
01:84